src/vmm/src/device_manager/resources.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::cell::RefCell;
+
+pub use vm_allocator::AllocPolicy;
+use vm_allocator::{AddressAllocator, IdAllocator};
+
+use crate::arch;
+
+/// A resource manager for (de)allocating interrupt lines (GSIs) and guest memory
+///
+/// At the moment, we support:
+///
+/// * GSIs for legacy x86_64 devices
+/// * GSIs for MMIO devicecs
+/// * Memory allocations in the MMIO address space
+#[derive(Debug)]
+pub struct ResourceAllocator {
+    // Allocator for device interrupt lines
+    gsi_allocator: RefCell<IdAllocator>,
+    // Allocator for memory in the MMIO address space
+    mmio_memory: RefCell<AddressAllocator>,
+}
```

Comment:
Note: I highlighted this code segment, but this comment is about the set of changes in this PR.

I don't like this design and here is why:
- Combining allocators for different resources into 1 struct seems unnecessary, especially considerring that in the end this sturct has different methods to allocate resources from each allocator separately.
- Sharing allocators between `Vmm` and `MMIODeviceManager` is counter intuitive. `Vmm` should not allocate anything, it has `MMIODeviceManager` for this.
- Making `Vmm` as an entity to create acpi tables is also unnecessary mixing of responsibilities. As I can see to create acpi tables, only `guest_memory` and `mmio_infos` are needed. It is simper to just pass them to some acpi table function that will do all the work and leave `Vmm` out of this.
- Spreading `Vmm` impementation between files (not just in `acpi/mod.rs` but also in `acpi/x86.rs`) degrades the code quality. It makes it harder to understand what is `Vmm` actually should do because you need to jump though the hoops to find all `imps` for it. Also as I can see, only `setup_arch_dsdt` method actually uses `self`, but then `pio_device_manager` impl for `append_aml_bytes` retuns a constant data, so this whole implementation for `Vmm` actually does not need `Vmm`.

I would propose something like this instead:
```rust
/// leave all allocators as they are now

/// Make use of the `MMIODeviceInfo` struct that is returned 
/// by the `vmm.mmio_device_manager.register_mmio_virtio_for_boot`
/// call. (We currently just drop the return value)
pub fn attach_virtio_device(...) -> Result<MMIODeviceInfo, ...> {
   ...
    vmm.mmio_device_manager
        .register_mmio_virtio_for_boot(vmm.vm.fd(), id, device)
        .map_err(RegisterMmioDevice)
        /// .map(|_| ()) <- just remove this line
}
/// Update all `attach_boot_timer_device`, `attach_entropy_device` and so on to also
/// return `MMIODeviceInfo`

pub fn build_microvm_for_boot(...)  {
   ...
    let mut mmio_device_infos = vec![];

    if vm_resources.boot_timer {
        let mmio_info = attach_boot_timer_device(&mut vmm, request_ts)?;
        mmio_device_infos.append(mmio_info);
    }

    if let Some(balloon) = vm_resources.balloon.get() {
        let mmio_info = attach_balloon_device(&mut vmm, balloon, event_manager)?;
        mmio_device_infos.append(mmio_info);
    }

    // Maybe worth unwrapping this into a loop which attaches
   // each block separately
    let mmio_infos = attach_block_devices(
        &mut vmm,
        &mut boot_cmdline,
        vm_resources.block.devices.iter(),
        event_manager,
    )?;
    mmio_device_infos.extend(&mmio_infos);
 
    // Same here
    let mmio_infos = attach_net_devices(&mut vmm, vm_resources.net_builder.iter(), event_manager)?;
    mmio_device_infos.extend(&mmio_infos);

    if let Some(unix_vsock) = vm_resources.vsock.get() {
        let mmio_info = attach_unixsock_vsock_device(&mut vmm, unix_vsock, event_manager)?;
        mmio_device_infos.append(mmio_info);
    }

    if let Some(entropy) = vm_resources.entropy.get() {
        let mmio_info = attach_entropy_device(&mut vmm, entropy, event_manager)?;
        mmio_device_infos.append(mmio_info);
    }

   ...

     configure_system_for_boot(
        &mut vmm,
        ...
        mmio_device_infos
    )?;
   ...
}

pub fn configure_system_for_boot(..., mmio_device_infos) {
  ...
  #[cfg(target_arch = "x86_64")]
  acpi_table::x86::create_acpi_table(&vmm.guest_memory, mmio_device_infos, vcpus.len())?;
  ...
}

// in x86/acpi module
pub fn create_acpi_table(guest_memory: &GuestMemory, mmio_device_infos: ..., vcpu_num: u32) -> Result<...> {
  let acpi_memory = AddressAllocator::new(...);
  let dsdt_addr = acpi_memory.allocate(...);
  let dsdt_table = Dsdt::new(&mmio_device_infos);
  dsdt_table.write_to_guest(&guest_memory, dsdt_addr);
  // repeat for all other tables
  ...
}

```